libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,863 @@
+       */
+      iterator end() {
+         auto prefix = make_prefix(table_name, name);
+         uint32_t itr = internal_use_do_not_use::kv_it_create(db, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_move_to_end(itr);
+
+         return {contract_name, itr, static_cast<kv_it_stat>(itr_stat), this};
+      }
+
+      /**
+       * Returns an iterator to the object with the lowest key (by this index) in the table.
+       * @ingroup keyvalue
+       *
+       * @return An iterator to the object with the lowest key (by this index) in the table.
+       */
+      iterator begin() {
+         auto prefix = make_prefix(table_name, name);
+         uint32_t itr = internal_use_do_not_use::kv_it_create(db, contract_name.value, prefix.data(), prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, "", 0);
+
+         return {contract_name, itr, static_cast<kv_it_stat>(itr_stat), this};
+      }
+
+      /**
+       * Returns a vector of objects that fall between the specifed range. The range is inclusive, exclusive.
+       * @ingroup keyvalue
+       *
+       * @tparam K - The type of the key. This will be auto-deduced by the key param.
+       *
+       * @param begin - The beginning of the range.
+       * @param end - The end of the range.
+       * @return A vector containing all the objects that fall between the range.
+       */
+      template <typename K>
+      std::vector<T> range(K begin, K end) {
+         auto begin_itr = lower_bound(begin);
+         auto end_itr = upper_bound(end);
+
+         bool include_last = find(end) != end_itr;
+
+         if (begin_itr == end_itr || begin_itr > end_itr) {
+            return {};
+         }
+
+         std::vector<T> return_values;
+
+         iterator itr = begin_itr;
+         while(itr != end_itr) {
+            return_values.push_back(itr.value());
+            ++itr;
+         }
+
+         if (include_last) {
+            return_values.push_back(itr.value());
+         }
+
+         return return_values;
+      }
+
+      key_type get_key(const T& inst) const { return key_function(inst); }
+
+   private:
+      std::function<key_type(const T&)> key_function;
+   };
+
+   /**
+    * @ingroup keyvalue
+    *
+    * @brief Defines a deleted index on an EOSIO Key Value Table
+    * @details Due to the way indexes are named, when deleting an index a "placeholder" index needs to be created instead.
+    * A null_kv_index should be created in this case. If using DEFINE_TABLE, just passing in nullptr will handle this.
+    */
+   class null_kv_index : public kv_index {};
+
+   /**
+    * Puts a value into the table. If the value already exists, it updates the existing entry.
+    * The key is determined from the defined primary index.
+    * @ingroup keyvalue
+    *
+    * @param value - The entry to be stored in the table.
+    */
+   void put(const T& value) {
```

Comment:
This doesn't clean up secondary indexes if the value already exists and secondary keys changed

Reply:
The best solution I could come up with is just deleting all secondary indexes before `set`ting them because I couldn't think of a way to derive the old value of the index (if it has changed) with the context we have here.

I'm open to a better solution if there is one.

Reply:
This PR was getting large, so I added the changes for this in a new PR https://github.com/EOSIO/eosio.cdt/pull/807 that goes into this branch.